Cython/Compiler/Nodes.py

```diff
@@ -1242,11 +1247,11 @@ def _analyse_template_types(self, env, base_type):
                     error(template_node.pos, "unknown type in template argument")
                     ttype = error_type
                 # For Python generics we can be a bit more flexible and allow None.
-            elif require_python_types and not ttype.is_pyobject:
+            elif check_require_types and not check_require_types(ttype):
                 if ttype.equivalent_type and not template_node.as_cython_attribute():
                     ttype = ttype.equivalent_type
                 else:
-                    error(template_node.pos, "%s[...] cannot be applied to non-Python type %s" % (
+                    error(template_node.pos, "%s[...] cannot be applied to type %s" % (
```

Comment:
I'm not sure if we need to change this error message. The fact that we mention "non-Python" types does not mean that _all_ non-Python types would reject `Optional[]`. I'm torn, though, and admit that there is a certain bit of suggestion here. Maybe @da-woods has a stronger opinion.

Reply:
I don't really have a strong opinion. Given you've changed the error message I think leave it changed.